docs: Added v1.0.0 information to upgrade guide (#6560)
aws/karpenter-provider-aws: website/content/en/preview/upgrading/upgrade-guide.md

WHEN CREATING A NEW SECTION OF THE UPGRADE GUIDANCE FOR NEWER VERSIONS, ENSURE THAT YOU COPY THE BETA API ALERT SECTION FROM THE LAST RELEASE TO PROPERLY WARN USERS OF THE RISK OF UPGRADING WITHOUT GOING TO 0.32.x FIRST
-->

> ### Upgrading to `1.0.0`+

Reviewer: I don't see us providing any detail on the API changes that were made in v1. Do we want to make a callout about the specific parts of the API that changed so that users can update their manifests that they may have written locally

Author: With the new "v1 Migration Guide" I kept the "Changelog" section of the previous migration guide. So far, I have just dropped in the API changes that I had described earlier based on when then needed to be included: before v1.0 upgrade or before v1.1 upgrade. Maybe we should discuss if we want to cover this content in a different way.

> ### Upgrading to `1.0.0`+

Reviewer: We didn't mention the name change for `karpenter.sh/disruption=disrupting` to `karpenter.sh/disrupted` in here

> ### Upgrading to `1.0.0`+

Author: Done.

{{% alert title="Warning" color="warning" %}}
`1.0.0`+ supports Karpenter v1 and v1beta1 APIs and will not work with earlier Provisioner, AWSNodeTemplate or Machine alpha APIs. Do not upgrade to `1.0.0`+ without first [upgrading to `0.32.x`]({{<ref "#upgrading-to-0320" >}}) or later. This version supports both the v1beta1 and v1 APIs, allowing you to migrate all of your existing APIs to v1beta1 APIs without experiencing downtime.
{{% /alert %}}

Before you begin upgrading to `1.0.0`, you should know that:

* Every Karpenter upgrade from pre-v1.0.0 versions must go through an upgrade to minor version `v1.0.0`.
* You must be on a version of Karpenter that supports NodePools, NodeClaims, and NodeClasses (`0.32.0`+ Karpenter versions support v1beta1 APIs).
* You must complete all `Before upgrading to 1.0.0` Prerequisites.
* You must complete all `Before upgrading to `1.1.0` Prerequisites before upgrading to `1.1.0` later. Although v1beta1 yaml files should work for `v1.0.0`, the will fail with the release of `v1.1.0`.

> #### Prerequisites

Reviewer 2: We need to call out that customer will need to update their IAM permissions when upgrading to 1.0

Consider two different categories of `1.0.0` prerequisites: those you must do before `1.0.0` upgrades and those you must do before `1.1.0` upgrades.

**Before upgrading to `1.0.0`**:

> * **Deprecated annotations, labels and tags are removed for v1.0.0**: For v1, `karpenter.sh/do-not-consolidate` (annotation), `karpenter.sh/do-not-evict
> (annotation)`, and `karpenter.sh/managed-by` (tag) all have support removed.

Reviewer: ```suggestion
* **Deprecated annotations, labels and tags are removed for v1.0.0**: For v1, `karpenter.sh/do-not-consolidate` (annotation), `karpenter.sh/do-not-evict` (annotation), and `karpenter.sh/managed-by` (tag) all have support removed.
```

Looks like there is an extra line that we may want to remove as well

The `karpenter.sh/managed-by`, which currently stores the cluster name in its value, is replaced by `eks:eks-cluster-name`, to allow
for [EKS Pod Identity ABAC policies](https://docs.aws.amazon.com/eks/latest/userguide/pod-id-abac.html).

> * **Zap logging removed**: Support for setting the Zap logging config was deprecated in beta and is now removed for v1. View the [Logging Configuration Section of the v1beta1 Migration Guide]({{<ref "../../v0.32/upgrading/v1beta1-migration#logging-configuration-is-no-longer-dynamic" >}}) for more details.

Reviewer: We may want to mention that we have retained support for `LOG_LEVEL`, `LOG_OUTPUT_PATHS` and `LOG_ERROR_OUTPUT_PATHS`

> * **metadataOptions could break workloads**: If you have workload pods that are not using `hostNetworking`, the updated default `metadataOptions`

Reviewer: I think we should be more specific about this or ref-out to docs on IMDS and metadata options. This sounds like this is going to be more common than it probably actually will be. Realistically, users that are affected by this are going to be ones that have pods that are accessing IMDS but are not running in hostNetworking mode. It's also worth noting that users can configure this to override this -- we just updated the default

* **Support for native Ubuntu AMI selection is dropped**: If you are using Ubuntu, be aware that Karpenter v1 no longer natively supports Ubuntu. To continue using Ubuntu in v1, you can update `amiSelectorTerms` in your EC2NodeClasses to identify Ubuntu as the AMI you want to use. See [reimplement amiFamily](https://github.com/aws/karpenter-provider-aws/pull/6569) for an example. Once you have done that, you can leave the amiFamily as Ubuntu and proceed to upgrading to v1. This will result in the following Transformation:
   ```yaml
   version: karpenter.k8s.aws/v1beta1
   kind: EC2NodeClass
   spec:
     amiFamily: Ubuntu
     amiSelectorTerms:
     - id: ami-foo
   ---
   # Conversion Webhook Output
   version: karpenter.k8s.aws/v1
   kind: EC2NodeClass
   metadata:
     annotations:
       compatibility.karpenter.k8s.aws/v1beta1-ubuntu: amiFamily,blockDeviceMappings
   spec:
     amiFamily: AL2
     amiSelectorTerms:
     - id: ami-foo
     blockDeviceMappings:
     - deviceName: 'dev/sda1'
       rootVolume: true
       ebs:
         encrypted: true
         volumeType: gp3
         volumeSize: 20Gi
   ```

>    **NOTE**: If you do not specify amiSelectorTerms before upgrading with an Ubuntu AMI, the conversion webhoot will fail. If you do this, downgrade to Karpenter `v0.37.1 and add amiSelectorTerms before upgrading again.

Reviewer: ```suggestion
   **NOTE**: If you do not specify amiSelectorTerms before upgrading with an Ubuntu AMI, the conversion webhook will fail. If you do this, downgrade to Karpenter `v0.37.1 and add amiSelectorTerms before upgrading again.
```

>    **NOTE**: If you do not specify amiSelectorTerms before upgrading with an Ubuntu AMI, the conversion webhoot will fail. If you do this, downgrade to Karpenter `v0.37.1 and add amiSelectorTerms before upgrading again.

Reviewer: ```suggestion
   **NOTE**: If you do not specify amiSelectorTerms before upgrading with an Ubuntu AMI, the conversion webhoot will fail. If you do this, update your EC2NodeClass to specify amiSelectorTerms on v1beta1, resolving the conversion webhook failures.
```

I don't think that we actually need to recommend that users downgrade when this happens. Realistically, they can just update the v1beta1 version of the NodeClass and everything should just work from there on out

* **Remove BootstrapMode annotation**: Karpenter will crash if NodePool resources contain the `BootstrapMode` annotation.
This annotation is no longer being added. If you are migrating an Ubuntu NodeClass to v1, you need to remove the `karpenter.k8s.aws/v1beta1-ubuntu` annotation.

* **KubeletConfiguration**: If you have multiple NodePools pointing to the same EC2NodeClass that have different kubeletConfigurations,
then you have to manually intervene to add more EC2NodeClasses and point their NodePools to them.
Otherwise, this will induce drift and you will have to roll your cluster.
If you have multiple NodePools pointing to the same EC2NodeClass, but they have the same configuration, then you can proceed with the migration
without having drift or having any additional NodePools or EC2NodeClasses configured.

* **amiSelectorTerms and amiFamily**: For v1, `amiFamily` is no longer required if you instead specify an `alias` in `amiSelectorTerms` in your `EC2NodeClass`. However, you need to update those settings if you are using:
   * A Custom amiFamily. You must ensure that the node is registered with the `karpenter.sh/unregistered` taint.
   * An Ubuntu AMI, as described earlier.

> #### Updated metrics

Reviewer: We made some changes to environment variables. Should we call those out in a section here?

| Disruption | karpenter_disruption_budgets_allowed_disruptions | karpenter_nodepools_allowed_disruptions |
| Disruption | karpenter_disruption_actions_performed_total | karpenter_voluntary_disruption_decisions_total |
| Provisioner | karpenter_provisioner_scheduling_simulation_duration_seconds | karpenter_scheduler_scheduling_duration_seconds |
| Provisioner | karpenter_provisioner_scheduling_queue_depth | karpenter_scheduler_queue_depth |
| Interruption | karpenter_interruption_received_messages | karpenter_interruption_received_messages_total |
| Interruption | karpenter_interruption_deleted_messages | karpenter_interruption_deleted_messages_total |
| Interruption | karpenter_interruption_message_latency_time_seconds | karpenter_interruption_message_queue_duration_seconds |
| NodePool     | karpenter_nodepool_usage | karpenter_nodepools_usage |
| NodePool     | karpenter_nodepool_limit | karpenter_nodepools_limit |
| NodeClaim    | karpenter_nodeclaims_terminated | karpenter_nodeclaims_terminated_total |
| NodeClaim    | karpenter_nodeclaims_disrupted | karpenter_nodeclaims_disrupted_total |
| NodeClaim    | karpenter_nodeclaims_created | karpenter_nodeclaims_created_total |

> This table shows v1beta1 metrics that were dropped for v1:

Reviewer: Since we dropped them, I think we should tell users the metrics that they are replaced with, if any. Happy to help provide that info so that we give the right guidance to everyone

| Metric type | Metric dropped for v1 |
| Disruption  | karpenter_disruption_replacement_nodeclaim_initialized_seconds |
| Disruption  | karpenter_disruption_queue_depth |
| Disruption  | karpenter_disruption_pods_disrupted_total |
|             | karpenter_consistency_errors |
| NodeClaim   | karpenter_nodeclaims_registered |
| NodeClaim   | karpenter_nodeclaims_launched |
| NodeClaim   | karpenter_nodeclaims_initialized |
| NodeClaim   | karpenter_nodeclaims_drifted |
| Provisioner | karpenter_provisioner_scheduling_duration_seconds |
| Interruption | karpenter_interruption_actions_performed |

#### Results

> When you upgrade to `1.0.0` CRDs, Karpenter supports serving both the `v1beta1` versions and the `v1` versions of NodePools, NodeClaims, and EC2NodeClasses.

Reviewer: Should we ref out to the K8s docs on what serving versions mean as well as what versions for CRDs generally mean, in case users aren't familiar

WHEN CREATING A NEW SECTION OF THE UPGRADE GUIDANCE FOR NEWER VERSIONS, ENSURE THAT YOU COPY THE BETA API ALERT SECTION FROM THE LAST RELEASE TO PROPERLY WARN USERS OF THE RISK OF UPGRADING WITHOUT GOING TO 0.32.x FIRST
-->

### Upgrading to `1.0.0`+

{{% alert title="Warning" color="warning" %}}
Karpenter `1.0.0` introduces v1 APIs, including _significant_ changes to the API and upgrade procedures for the Karpenter controllers. **Do not** upgrade to `1.0.0`+ without referencing the [v1 Migration Upgrade Procedure]({{<ref "v1-migration#upgrade-procedure" >}}).

> This version includes **dual support** for both v1 and v1beta1 APIs to ensure that you can slowly migrate your existing NodePool, EC2NodeClass, and NodeClaim v1beta1 APIs to v1 APIs.

Author: Done.

WHEN CREATING A NEW SECTION OF THE UPGRADE GUIDANCE FOR NEWER VERSIONS, ENSURE THAT YOU COPY THE BETA API ALERT SECTION FROM THE LAST RELEASE TO PROPERLY WARN USERS OF THE RISK OF UPGRADING WITHOUT GOING TO 0.32.x FIRST
-->

### Upgrading to `1.0.0`+

{{% alert title="Warning" color="warning" %}}
`1.0.0`+ supports Karpenter v1 and v1beta1 APIs and will not work with earlier Provisioner, AWSNodeTemplate or Machine alpha APIs. Do not upgrade to `1.0.0`+ without first [upgrading to `0.32.x`]({{<ref "#upgrading-to-0320" >}}) or later. This version supports both the v1beta1 and v1 APIs, allowing you to migrate all of your existing APIs to v1beta1 APIs without experiencing downtime.
{{% /alert %}}

Before you begin upgrading to `1.0.0`, you should know that:

* Every Karpenter upgrade from pre-v1.0.0 versions must go through an upgrade to minor version `v1.0.0`.

> * You must be on a version of Karpenter that supports NodePools, NodeClaims, and NodeClasses (`0.33.0`+ Karpenter versions support v1beta1 APIs).

Author: Done.

* You must complete all `Before upgrading to 1.0.0` Prerequisites.

> * Support for v1beta1 yaml files will end with the release of `v1.1.0`.

Author: Okay.

> * You must complete all `Before upgrading to `1.1.0` Prerequisites before upgrading to `1.1.0` later. Although v1beta1 yaml files should work for `v1.0.0`, the will fail with the release of `v1.1.0`.

Reviewer 3: I wonder if we need to include this yet, since we won't have this included until when we actually cut the release.

Reviewer: +1, though we shouldn't forget to come back and include it. Maybe a PR that we throw in draft while we are merging this v1 PR?

Consider two different categories of `1.0.0` prerequisites: those you must do before `1.0.0` upgrades and those you must do before `1.1.0` upgrades.

Before upgrading to `1.0.0`:

* **Deprecated annotations, labels and tags are removed for v1.0.0**: For v1, `karpenter.sh/do-not-consolidate` (annotation), `karpenter.sh/do-not-evict
(annotation)`, and `karpenter.sh/managed-by` (tag) all have support removed.
The `karpenter.sh/managed-by`, which currently stores the cluster name in its value, is replaced by `eks:eks-cluster-name`, to allow
for [EKS Pod Identity ABAC policies](https://docs.aws.amazon.com/eks/latest/userguide/pod-id-abac.html).

* **Zap logging removed**: Support for setting the Zap logging config was deprecated in beta and is now removed for v1. View the [Logging Configuration Section of the v1beta1 Migration Guide]({{<ref "../../v0.32/upgrading/v1beta1-migration#logging-configuration-is-no-longer-dynamic" >}}) for more details.

* **metadataOptions could break workloads**: If you have workload pods that are not using `hostNetworking`, the updated default `metadataOptions`
could cause your containers to break when you apply new EC2NodeClasses on v1.

> * **Ubuntu alias dropped**: If you are using Ubuntu, be aware that Karpenter v1 no longer supports an `alias` for using Ubuntu.

Author: Done.

> * **Ubuntu alias dropped**: If you are using Ubuntu, be aware that Karpenter v1 no longer supports an `alias` for using Ubuntu.

Reviewer 3: More details on this: 
"If you're setting AMIFamily: Ubuntu" and have no AMISelectorTerms defined, Karpenter automatically upgrades AMIs for you. Automatic upgrades are not supported for Ubuntu in v1, and you must pin your AMIs before upgrading"

* **Migrating userData**: Consider migrating your userData ahead of the v1.0.0 upgrade. See the description of `AMISelectorTerms` below.

Before upgrading to `1.1.0` (though okay to do for for `1.0.0`):

> * **v1beta1 support gone`**: Migrate all Karpenter yaml files ([NodePools]({{<ref "../concepts/nodepools" >}}), [EC2NodeClasses]({{<ref "../concepts/nodeclasses" >}}), and so on) to v1.

Reviewer 4: We should probably specify that it's not just files that you apply that need to be migrated, but all resources in the cluster also need to be on v1. It's possible (but probably pretty unlikely) that some resources still may be stored as v1beta1 in ETCD if no writes had been made to them since the v1 upgrade. Users could use a tool like the [kube-storage-version-migrator](https://github.com/kubernetes-sigs/kube-storage-version-migrator) to handle this.

Author: Done.

In `1.1.0`, v1beta1 is not supported.

* **Remove kubelet objects from NodePools**: Check that NodePools no longer contain the kubelet-configuration annotation (`karpenter.sh/v1beta1-kubelet-conversion` annotation).
Karpenter will crash if NodePool resources contain this annotation.

> * **Remove BootstrapMode annotation**: Check that EC2NodeClasses no longer contain the `BootstrapMode` annotation.

Author: Done.

> * **Remove BootstrapMode annotation**: Check that EC2NodeClasses no longer contain the `BootstrapMode` annotation.

Reviewer 3: This is stale, we don't have a bootstrap mode annotation anymore. I think you can just remove this line altogether.

Karpenter will crash if NodePool resources contain this annotation.

* **KubeletConfiguration**: If you have multiple NodePools pointing to the same EC2NodeClass that have different kubeletConfigurations,
then you have to manually intervene to add more EC2NodeClasses and point their NodePools to them.

> Otherwise, this will induce drift.

Author: Okay.

If you have multiple NodePools pointing to the same EC2NodeClass, but they have the same configuration, then you can proceed with the migration

> without having drift or having any additional NodePools or EC2NodeClasses configured.

Reviewer 2: I'm not sure this is true. When users makes an update to their nodeclassRef they will be drifted

Reviewer 2: I think we can combine to the two cases above

* **AMISelectorTerms**: If you are using the Custom AMIFamily or are using an AMIFamily with no amiSelectorTerms, then you don’t need to take
any action on your userData for this migration.
If you are using an AMIFamily with amiSelectorTerms, then you are relying on the merge logic between Karpenters' known userData for an AMIFamily
and your custom userData.
This is no longer be possible with v1. You need to update your userData to match what is produced when Karpenter does this merge,

> because this will not be auto-generated for you.

Author: Please check to make sure that I added this information correctly.
